dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/ProgramRuleDeletionHandler.java

```diff
@@ -72,4 +78,31 @@ public void deleteProgram( Program program )
             programRuleService.deleteProgramRule( programRule );
         }
     }
+
+    @Override
+    public String allowDeleteProgramStage( ProgramStage programStage )
+    {
+        List<ProgramRule> programRuleVariables = programRuleService
+            .getProgramRule( programStage.getProgram() )
+            .stream()
+            .filter( pr -> isLinkedToProgramStage( programStage, pr ) )
+            .collect( Collectors.toList() );
+
+        if ( programRuleVariables.isEmpty() )
+        {
+            return null;
+        }
+
+        return programRuleVariables.stream()
+            .map( BaseIdentifiableObject::getName )
+            .collect( Collectors.joining( ", " ) );
+    }
```

Comment:
Minor: Might be slightly more compact and avoid to stream twice, like this:
 
```suggestion
    @Override
    public String allowDeleteProgramStage( ProgramStage programStage )
    {
        String programRuleVariables = programRuleService
            .getProgramRule( programStage.getProgram() )
            .stream()
            .filter( pr -> isLinkedToProgramStage( programStage, pr ) )
            .map( BaseIdentifiableObject::getName )
            .collect( Collectors.joining( ", " ) );

        return StringUtils.isBlank( programRuleVariables ) ? null : programRuleVariables;
    }
```